### PR TITLE
improve derivation of `BFieldCodec` for structs with only fixed-length fields

### LINKED ISSUE #124: Error in derived implementation of `BFieldCodec` for vectors
Minimal example

```rust
use twenty_first::shared_math::tip5::Digest;

#[derive(Clone, Copy, Debug, Serialize, Deserialize, PartialEq, Eq, Hash, GetSize, BFieldCodec)]
pub struct Foo {
    pub bar: Digest,
}

#[cfg(test)]
mod tests {
    use rand::random;
    use super::*;

    #[test]
    fn foo_vec_test() {
        for i in 0..5 {
            let foos = vec![
                Foo {
                    bar: random(),
                };
                i
            ];
            let encoded = foos.encode();
            let decoded = *Vec::<Foo>::decode(&encoded).unwrap();
            assert_eq!(foos, decoded);
        }
    }

```

Error message:
```
thread 'foo_vec_test' panicked at 'called `Result::unwrap()` on an `Err` value: Length indication plus one must match actual sequence length. Claimed vector length was 1. Item size was 5. Sequence length was 7.', ...
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

The encoding prepends the size of each element, but the decoder does not expect that.

### PR DESCRIPTION
**Pull request by jan-ferdinand**
Fix #124

### CODE REVIEW

File: twenty-first/src/shared_math/bfield_codec.rs
Comment (jan-ferdinand) [inline twenty-first/src/shared_math/bfield_codec.rs:706]: I'm not super happy about this. The underlying requirement comes from initiating the result vector as `Vec::with_capacity(num_elements)` in `impl<T> BFieldCodec for Vec<T>::decode()`, which is nice to have but apparently introduces a DOS vector. The alternative is to use `vec![]` and bite the cost of internal re-allocations.

Thought, @Sword-Smith?

Comment (jan-ferdinand) [reply twenty-first/src/shared_math/bfield_codec.rs]: After writing above comment, I feel much more strongly inclined to use `vec![]` instead of `Vec::with_capacity(random_persons_claim)`.

**Review by Sword-Smith — CHANGES_REQUESTED** [review]
This certainly works in isolation, and it might be an improvement.

A significant change, though, is that the encoding of a `struct` no longer prepends the length of each field. This will break things downstream, especially things in https://github.com/TritonVM/tasm-lib/tree/asz/removal-record-integrity/src/structure.

It might be advantageous to always prepend the field length when encoding a struct, even if it is known at compile-time as that makes any TASM program that deals with these encodings simpler.

I think @aszepieniec should weigh in on this.

This new rule that @jan-ferdinand has programmed would mean that the field size is *not* prepended if it is known statically, otherwise the field size *is* prepended. It makes for shorter encodings, but also for potentially more complicated TASM programs to get a specific field from a struct.

**Comment by jan-ferdinand** [comment]
Good point. I wasn't aware of such dependencies downstream. In isolation, it seems to me that my suggested solution is a rather clean approach. My hope is that the approach can be transferred to the TASM context as well.

**Comment by aszepieniec** [comment]
If every field is length-prepended, then the procedure for "get field x" can be struct-agnostic simply by indexing and hopping over fields until a decrementing counter reaches zero. But if some fields are not length-prepended, then the "get field x" method has to be a custom method for every struct, for every field.

I started implementing a collection of "get field x" methods for `TransactionKernel` assuming (wrongly) that statically-sized fields are not length-prepended. The result was a monster. I'm very happy to use the generic "get nth field" method that works for any struct under the length-prepended encoding and I would strongly suggest we not change this property.

**Comment by Sword-Smith** [comment]
After talking to @aszepieniec we have agreed that this proposal from @jan-ferdinand is the way we want to go. But please don't merge before conferring with @aszepieniec or me, as there will be downstream complications.

**Comment by Sword-Smith** [comment]
@aszepieniec and myself should go over this new `BFieldCodec` derivation before merging, since we have the most experience working with these encodings in TASM.

- [ ] Approvable by @aszepieniec 
- [ ] Approvable by @Sword-Smith 
- [ ] Approvable by @jan-ferdinand (optional)

**Review by Sword-Smith — CHANGES_REQUESTED** [review]
No merge before approved by requested reviewers

**Comment by aszepieniec** [comment]
Approved via #136

**Pull request closed without merging**
